Add support for generating a full `ColorScheme` from a single color

darrenaustin opened the pull request:
As part of the [migration to Material 3](https://github.com/flutter/flutter/issues/91605), this PR adds a new APIs to make it easy to generate new Material 3 ColorSchemes from just a single color.

The [design doc](https://flutter.dev/go/colorscheme-generation) discusses this in more detail.

The new APIs are as follows:

### `ColorScheme.fromSeed()`

You can construct a complete Material 3 ColorScheme derived from the tones of a single seed color with:

```dart
final lightScheme = ColorScheme.fromSeed(seedColor: Colors.green);
final darkScheme = ColorScheme.fromSeed(seedColor: Colors.green, brightness: Brightness.dark);
```

### `ThemeData(Color colorSchemeSeed, ...)`

There is a new `colorSchemeSeed` parameter to the `ThemeData` factory constructor that will allow the app to generate the theme's color scheme from a seed color:

```dart
final lightTheme = ThemeData(colorSchemeSeed: Colors.orange, ...);
final darkTheme = ThemeData(colorSchemeSees: Colors.orange, brightness: Brightness.dark, ...);
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

darrenaustin commented:
This will require #93427 to land first as we need the updated colorscheme.

Note to reviewers, I broke this up into multiple commits to make it easier to review the addition of the dependency on `material_color_utiltilies` (https://github.com/flutter/flutter/pull/93463/commits/d41ffed0cb98597e4bc14d018b3abbb1042a9859)  and the new APIs (https://github.com/flutter/flutter/pull/93463/commits/478229b706f8d684ee4a43a820914ded57f59b75).

rami-a reviewed (approved):
LGTM

guidezpl commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -200,8 +200,17 @@ enum MaterialTapTargetSize {
+  /// This pair of values is used to construct the [colorScheme] by using
+  /// [ColorScheme.fromSeed].
+  ///
+  /// For historical reasons, instead of using a [colorSchemeSeed], you can
```
Good to note this, also might be worth noting the primary alternative is to use `ColorScheme.fromSeed`,  customize as needed with `copyWith` and set `colorScheme`

darrenaustin replied:
Good point. I updated the comment with something about that.

guidezpl commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -236,6 +246,7 @@ class ThemeData with Diagnosticable {
     // properties (as well as brightness, primaryColorBrightness, and primarySwatch)
     // will gradually be phased out, see https://github.com/flutter/flutter/issues/91772.
     ColorScheme? colorScheme,
+    Color? colorSchemeSeed,
```
nit: reorder to put `colorSchemeSeed` and `brightness` above `colorScheme` since it's the recommended path

darrenaustin replied:
But in this case the important one is the `colorScheme` the others are just used to generate one, so I think it makes more sense to have it first and the related non-property parameters after it.

guidezpl commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -236,6 +246,7 @@ class ThemeData with Diagnosticable {
     // properties (as well as brightness, primaryColorBrightness, and primarySwatch)
```
Please update all instances of this comment

darrenaustin replied:
I updated this one, but the other similar comments didn't reference these parameters, so I left them alone.

guidezpl commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -390,8 +400,35 @@ class ThemeData with Diagnosticable {
+    if (colorSchemeSeed != null) {
+      colorScheme = ColorScheme.fromSeed(seedColor: colorSchemeSeed, brightness: _brightness);
+
+      // For surfaces that use primary color in light themes and surface color in dark
+      final Color primarySurfaceColor = isDark ? colorScheme.surface : colorScheme.primary;
+      final Color onPrimarySurfaceColor = isDark ? colorScheme.onSurface : colorScheme.onPrimary;
+
+      // Default some of the color settings to values from the color scheme
+      primaryColor = primarySurfaceColor;
+      primaryColorBrightness = ThemeData.estimateBrightnessForColor(primarySurfaceColor);
+      canvasColor ??= colorScheme.background;
+      accentColor ??= colorScheme.secondary;
+      accentColorBrightness ??= ThemeData.estimateBrightnessForColor(colorScheme.secondary);
+      scaffoldBackgroundColor ??= colorScheme.background;
+      bottomAppBarColor ??= colorScheme.surface;
+      cardColor ??= colorScheme.surface;
+      dividerColor ??= colorScheme.outline;
+      backgroundColor ??= colorScheme.background;
+      dialogBackgroundColor ??= colorScheme.background;
+      indicatorColor ??= onPrimarySurfaceColor;
+      errorColor ??= colorScheme.error;
+      applyElevationOverlayColor ??= isDark;
+    }
+    applyElevationOverlayColor ??= false;
```
A bit worried about this, adds complexity and impacts future deprecation of these color properties. Is this temporary?

If kept, requires documentation to be updated for all of these.

darrenaustin replied:
This is a temporary convenience to make sure that components that are still using these colors will have reasonable default values from the generated color scheme. This is similar to what we do in `ThemeData.from`.

HansMuller commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -14,6 +14,106 @@
 
 version: 1
 transforms:
+  # Changes made in https://github.com/flutter/flutter/pull/93427
```
These changes will need to be covered in a migration guide.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
 
-/// A set of twelve colors based on the
-/// [Material spec](https://material.io/design/color/the-color-system.html)
+/// A set of colors based on the
```
We might as well admit how many colors there are here: a set of 25 colors...

It would be helpful to explain the Foo and onFoo naming conventions and to provide an overview of the colors.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
+  ///
+  /// For the color parameters that are nullable, it is still recommended
+  /// that applications provide values for them. They are only nullable due
+  /// to backwards compatibility concerns.
```
The migration guide can point to the section in the design doc that explains what the backwards compatibility concerns are.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
+  ///
+  /// If a color is not provided, the closest fallback color from the given
+  /// colors will be used for it (e.g. [primaryContainer] will default
+  /// to [primary]). Material 3 makes use of these colors for many component
```
We need an reference to ThemeData.useMaterial3 here and a summary of the subset of colors the components will depend on if it's false.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
 @immutable
 class ColorScheme with Diagnosticable {
   /// Create a ColorScheme instance.
+  ///
```
I think it would be helpful to call out ColorScheme.fromSeed() in the 1st paragraph; since that's likely to be preferable to this constructor.

kwabenaG replied:
It makes sense 👍🏾

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
+       _primaryVariant = primaryVariant,
+       _secondaryVariant = secondaryVariant;
+
+  /// Generate a [ColorScheme] based around the given `seedColor`.
```
based around => based on (or derived from)

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
+  /// If any of the optional color parameters are non-null they will be
+  /// used in place of the generated colors for that field in the resulting
+  /// color scheme.
+  ///
```
Hopefully a fun interactive color-scheme generator example appears here...

darrenaustin replied:
It is on my list, but it would be a lot easier if we had a built in color picker... 😄

HansMuller replied:
You could always use 3 or 4 sliders for a rough and ready color picker.  Adding an example in a separate PR would be fine.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
 
   /// Create a ColorScheme based on a purple primary color that matches the
   /// [baseline Material color scheme](https://material.io/design/color/the-color-system.html#color-theme-creation).
   const ColorScheme.light({
```
As with the basic constructor, best to refer to ColorScheme.fromSeed() here.

darrenaustin replied:
Not sure I get why here. If they are looking for a pre-built 'light' color scheme, why would they want to use `fromSeed` (other than to make sure they get a full color scheme), and what should they use for the seed color? This kind of points out the problem with not having a M3 version of this and `dark`.

HansMuller replied:
I was thinking that it would be helpful to get apps off of ColorScheme.light(), since there isn't a perfect migration path that preserves the existing colors and adds new ones are comparable to what you'd get from ColorScheme.fromSeed().  Hopefully we can migrate developers that have used ColorScheme.light() to something like ColorScheme.fromSeed(seedColor: const Color(_just the right purple seed color here_). So for now we just remind them that the fromSeed factory is almost as trivial to write and so much more useful.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
 
   /// Create the recommended dark color scheme that matches the
   /// [baseline Material color scheme](https://material.io/design/color/dark-theme.html#ui-application).
   const ColorScheme.dark({
```
As with the basic constructor, best to refer to ColorScheme.fromSeed() here.

HansMuller commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -4,167 +4,508 @@
+
+  /// Generate a [ColorScheme] based around the given `seedColor`.
+  ///
+  /// Using the seedColor as a starting point, a set of tonal palettes are
```
Is there a way to generate a high-contrast color scheme from a seed color? It would help to provide a path from _all_ of the existing factories, to ColorScheme.fromSeed().

darrenaustin replied:
I didn't see an obvious way to do this with `material_color_utilities`, but @guidezpl would know.

guidezpl replied:
Sorry for the late reply.. There isn't

HansMuller reviewed (commented):
This all looks generally good to me. We're going to need a really good migration guide.

darrenaustin commented:
> This all looks generally good to me. We're going to need a really good migration guide.

Thanks for the feedback. I have addressed most of your comments with (https://github.com/flutter/flutter/pull/93463/commits/0b2d93ff612cc6e0c3e5eae1d7d1798553d27a06), let me know what you think.

A migration guide is next on my list, so I took notes from your comments here. Thx.

HansMuller reviewed (approved):
LGTM

You proposed breaking this into the package dependency and then the ColorScheme changes. That sounds like a good idea.

darrenaustin committed:
Added colorSchemeSeed parameter to the ThemeData constructor, and ColorScheme.fromSeed API.

darrenaustin commented:
I just landed the dependency changes for material_color_utilities with #94377, so this is just the new APIs.

guidezpl commented on packages/flutter/lib/src/material/color_scheme.dart:
```diff
@@ -97,6 +145,127 @@ class ColorScheme with Diagnosticable {
+    switch (brightness) {
+      case Brightness.light:
+        return ColorScheme(
+          primary: primary ?? Color(palette.primary.get(40)),
```
Since these numerical values are most likely to change (just did last week), I think it'd be idea for use to use `Scheme` and create the `ColorScheme` instance indirectly, etc. `primary: scheme.primary`. That way, no need to keep it in sync manually
